KAFKA-5501: introduce async ZookeeperClient

onurkaraman opened the pull request:
Synchronous zookeeper writes means that we wait an entire round trip before doing the next write. With respect to the controller, these synchronous writes are happening at a per-partition granularity in several places, so partition-heavy clusters suffer from the controller doing many sequential round trips to zookeeper.
- PartitionStateMachine.electLeaderForPartition updates leaderAndIsr in zookeeper on transition to OnlinePartition. This gets triggered per-partition sequentially with synchronous writes during controlled shutdown of the shutting down broker's replicas for which it is the leader.
- ReplicaStateMachine updates leaderAndIsr in zookeeper on transition to OfflineReplica when calling KafkaController.removeReplicaFromIsr. This gets triggered per-partition sequentially with synchronous writes for failed or controlled shutdown brokers.

onurkaraman committed:
use async zookeeper apis everywhere

Synchronous zookeeper writes means that we wait an entire round trip before doing the next write. With respect to the controller, these synchronous writes are happening at a per-partition granularity in several places, so partition-heavy clusters suffer from the controller doing many sequential round trips to zookeeper.
- PartitionStateMachine.electLeaderForPartition updates leaderAndIsr in zookeeper on transition to OnlinePartition. This gets triggered per-partition sequentially with synchronous writes during controlled shutdown of the shutting down broker's replicas for which it is the leader.
- ReplicaStateMachine updates leaderAndIsr in zookeeper on transition to OfflineReplica when calling KafkaController.removeReplicaFromIsr. This gets triggered per-partition sequentially with synchronous writes for failed or controlled shutdown brokers.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5686/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5672/
Test PASSed (JDK 8 and Scala 2.12).

junrao commented on core/src/main/scala/kafka/controller/ZookeeperClient.scala:
```diff
@@ -0,0 +1,202 @@
+    var state = sessionContext.get().getState
+    while (!state.isConnected && state.isAlive) {
+      isConnectedOrExpiredCondition.await()
+      state = sessionContext.get().getState
```
We need to verify if this works with sasl enabled in ZK client. I am not sure if zookeeper.getState() reflects KeeperState.SaslAuthenticated. Also, since isConnectedOrExpiredCondition is only notified through Watcher's state change event, it seems it's more nature to pick up state from the watcher instead of reading from zookeeper.getState().

junrao commented on core/src/main/scala/kafka/controller/ZookeeperClient.scala:
```diff
@@ -0,0 +1,202 @@
+          responseQueue.add(DeleteResponse(rc, path, ctx))
+          countDownLatch.countDown()
+        }}, ctx)
+      case ExistsRequest(path, ctx) => sessionContext.get().exists(path, false, new StatCallback {
```
It seems that it's useful to automatically set the watcher flag based on whether the path is registered through registerZNodeChangeHandler().

onurkaraman replied:
Can you elaborate?

The watcher flag is already set during the call to the handler registration apis. A later call with the watcher flag set to false does not cancel out an earlier call with the watcher flag set to true.

junrao replied:
Well, once the watcher fires, we need to reset the watcher again. It may not be convenient for the caller to explicitly call register again. Instead, it's more convenient if the watcher is set automatically on the read after the watcher fires, which is typically what the application has to do (This is what ZkClient does.).

junrao commented on core/src/main/scala/kafka/controller/ZookeeperClient.scala:
```diff
@@ -0,0 +1,202 @@
+
+  def registerZNodeChangeHandler(zNodeChangeHandler: ZNodeChangeHandler): Unit = {
+    zNodeChangeHandlers.put(zNodeChangeHandler.path, zNodeChangeHandler)
+    sessionContext.get().exists(zNodeChangeHandler.path, true)
```
For consistency, we probably want to call batch() so that we can consolidate on using the async api? Once we do that, would it be useful to make this a batch api so that we can register multiple watchers in a single call?

junrao commented on core/src/main/scala/kafka/controller/ZookeeperClient.scala:
```diff
@@ -0,0 +1,202 @@
+  @volatile private var stateChangeHandlerOpt: Option[StateChangeHandler] = None
+  @volatile private var zooKeeper = new ZooKeeper(connectString, sessionTimeout, ZookeeperClientWatcher)
+  private val sessionContext = new ThreadLocal[ZooKeeper] {
+    override def initialValue(): ZooKeeper = zooKeeper
```
I am not sure if we need ThreadLocal to store the zookeeper object. zookeeper api is in general thread-safe. So, the only case that we need to synchronize is when a new zookeeper object is created. This can probably be as simple as making the zookeeper object volatile.

onurkaraman replied:
I added the ThreadLocal `sessionContext` so that threads have the opportunity to react to expired sessions (and not accidentally send requests over the new session) even if another thread reinitialized the zookeeper session.

junrao replied:
Most ZK accesses are oblivious to which ZK session the access is on. The only exception is the controller logic and the registration of the broker id. So, we can just register a session expiration handler, in which we first clear the controller queue and add a new session expiration event to the controller queue. Then, once the controller thread picks up the session expiration event, it can create a new ZK session, re-register the broker id and do controller reelection. No other threads can create new ZK session. If they get an error on the current session, they can simply wait for the ZK state to be ready again and resume the operation.

onurkaraman replied:
I think my main concern with the above comment and a few others below is that they're very specific to the current usage.

junrao commented on core/src/main/scala/kafka/controller/ZookeeperClient.scala:
```diff
@@ -0,0 +1,202 @@
+          responseQueue.add(SetACLResponse(rc, path, ctx, stat))
+          countDownLatch.countDown()
+        }}, ctx)
+      case GetChildrenRequest(path, ctx) => sessionContext.get().getChildren(path, false, new Children2Callback {
```
Should we add the support for the multi request?

junrao commented on core/src/main/scala/kafka/controller/ZookeeperClient.scala:
```diff
@@ -0,0 +1,202 @@
+    batch(Seq(request)).head
+  }
+
+  def batch(requests: Seq[AsyncRequest]): Seq[AsyncResponse] = {
```
For consistency, perhaps this method can be named handle() as well? Also, it would be useful to add some comment that describes the behavior of the method.

onurkaraman replied:
Sounds good. I was thinking the same.

junrao commented on core/src/main/scala/kafka/controller/ZookeeperClient.scala:
```diff
@@ -0,0 +1,202 @@
+
+  def getState: States = sessionContext.get().getState
+
+  def waitUntilConnectedOrExpired: Boolean = inLock(isConnectedOrExpiredLock) {
```
Could we add a comment to describe the return value of this method?

onurkaraman replied:
The documentation is intentionally sparse and the class has no tests because we're still figuring out the API.

junrao commented on core/src/main/scala/kafka/controller/ZookeeperClient.scala:
```diff
@@ -0,0 +1,202 @@
+    zNodeChildChangeHandlers.remove(path)
+  }
+
+  def initialize(): Unit = inLock(initializationLock) {
```
This is supposed to be called after the client's ZK session expires, right? If so, this probably should be called sth like reconnect()?

onurkaraman replied:
"reconnect" would be misleading with respect to the terminology used for zookeeper client state:
{CONNECTING, ASSOCIATING, CONNECTED, CONNECTEDREADONLY, CLOSED, AUTH_FAILED, NOT_CONNECTED}

junrao replied:
newSession()?

junrao commented on core/src/main/scala/kafka/controller/ZookeeperClient.scala:
```diff
@@ -0,0 +1,202 @@
+  val path: String
+  def handleCreation: Unit
+  def handleDeletion: Unit
+  def handleDataChange: Unit
```
In ZkClient, the 3 independent events (create, delete and change) are consolidated into a single callback handleDataChange(), since in all those cases, the processing is likely going to be the same (read the current data and act on it). So, I am wondering if we should do the same thing here for simplicity.

onurkaraman replied:
This isn't quite true. IZkDataListener has two callbacks: handleDataChange and handleDataDeleted.

onurkaraman replied:
There are 4 Watcher.EventTypes in total:
NodeCreated
NodeDeleted
NodeDataChanged
NodeChildrenChanged

I took another look at org.apache.zookeeper.server.DataTree.

A data change watcher can return with types:
NodeCreated
NodeDeleted
NodeDataChanged

A child change watcher can return with types:
NodeDeleted
NodeChildrenChanged

A few interesting findings:
1. NodeDeleted applies to both data change watchers and child change watchers. My PR currently assumes NodeDeleted only applies to data change watchers.
2. Looking at ZkClient.processDataOrChildChange, I found somewhat of a mismatch between what the child change watcher gets called on and what IZkChildListener gets called on:
NodeCreated
NodeDeleted
NodeChildrenChanged

junrao commented on core/src/main/scala/kafka/controller/ZookeeperClient.scala:
```diff
@@ -0,0 +1,202 @@
+  }
+
+  def initialize(): Unit = inLock(initializationLock) {
+    if (!zooKeeper.getState.isAlive) {
```
Do we need to check the state? It seems that we can just blindly create a new zookeeper object.

onurkaraman replied:
We should only create a new session if the current session expired. If multiple threads call initialize, we don't want multiple sessions to be made.

junrao replied:
Hmm, but the check for zooKeeper.getState.isAlive is not going to be reliable since the state may not be alive immediately on a newly created session. Perhaps we can maintain a KeeperState that's updated through the watcher. Then, we can only create a new session if that state is in expiration state. After creating a new Zookeeper object, we can first set the state to Disconnected.

onurkaraman replied:
isAlive is defined as:
```
public boolean isAlive() {
    return this != CLOSED && this != AUTH_FAILED;
}
```
The initial state for Zookeeper is NOT_CONNECTED. This means that it actually is reliably defined as alive on a newly created session.

So I think this logic is correct as is.

junrao replied:
Ok, we can leave this as it is. However, if we need to maintain the KeeperState through watcher for other reasons, we probably want to consolidate the check on that KeeperState.

junrao reviewed (commented):
@onurkaraman Thanks for the patch. Left a few comments below. Also, could we add some unit tests?

onurkaraman committed:
invoke session initialization from the EventThread while blocking user thread

junrao commented on core/src/main/scala/kafka/controller/ZookeeperClient.scala:
```diff
@@ -0,0 +1,202 @@
+        inLock(isConnectedOrExpiredLock) {
+          isConnectedOrExpiredCondition.signalAll()
+        }
+        stateChangeHandlerOpt.foreach(_.handleStateChange)
```
For the state change handler, it seems that we only need to propagate the session expiration event?

junrao commented on core/src/main/scala/kafka/controller/ZookeeperClient.scala:
```diff
@@ -0,0 +1,202 @@
+    override def initialValue(): ZooKeeper = zooKeeper
+  }
+
+  def handle(request: AsyncRequest): AsyncResponse = {
```
It may also be convenient to have a blocking method that keeps retrying until the operation succeeds, instead of propagating connectionLossException back. Most ZK accesses can just use this method since they don't care about the ZK session the access is on and they just wan the operation to complete. For controller and broker registration, they will use the other api without the auto retry.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6092/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6077/
Test PASSed (JDK 8 and Scala 2.12).

junrao commented on core/src/main/scala/kafka/controller/ZookeeperClient.scala:
```diff
@@ -0,0 +1,214 @@
+          responseQueue.add(ExistsResponse(rc, path, ctx, stat))
+          countDownLatch.countDown()
+        }}, null)
+      case GetChildrenWithWatcherRequest(path) => zooKeeper.getChildren(path, true, new Children2Callback {
```
I had a comment on this earlier. Instead of having a separate GetChildrenWithWatcherRequest, it seems that it's simpler to just set the watcher flag based on whether the path is registered in zNodeChildChangeHandlers. In all our use cases, if a client registers a watcher on a path, the client would always want the watcher to be reregistered unless it explicitly calls unregisterZNodeChildChangeHandler().

junrao commented on core/src/main/scala/kafka/controller/ZookeeperClient.scala:
```diff
@@ -0,0 +1,214 @@
+    if (!zooKeeper.getState.isAlive) {
+      zNodeChangeHandlers.clear()
+      zNodeChildChangeHandlers.clear()
+      zooKeeper = new ZooKeeper(connectString, sessionTimeout, ZookeeperClientWatcher)
```
As described in KAFKA-5473, the instantiating of ZooKeeper can throw IOException (e.g., if the ZK server host can't be resolved). We need to handle this properly. One way is probably to backoff a bit and keep retrying initialize()?

junrao commented on core/src/main/scala/kafka/controller/ZookeeperClient.scala:
```diff
@@ -0,0 +1,214 @@
+case class GetChildrenRequest(path: String, ctx: Any) extends AsyncRequest
+
+sealed trait AsyncResponse
+case class CreateResponse(rc: Int, path: String, ctx: Any, name: String) extends AsyncResponse
```
For all the common fields like rc, path, ctx, perhaps we could pull them to AsyncResponse?

onurkaraman replied:
Will do.

junrao commented on core/src/main/scala/kafka/controller/ZookeeperClient.scala:
```diff
@@ -0,0 +1,214 @@
+      isConnectedOrExpiredCondition.await()
+      state = zooKeeper.getState
+    }
+    state.isConnected
```
If this method returns false, it doesn't always mean that the session has expired since it's possible for the state to be AUTH_FAILED. Perhaps we should return the state back and let the caller check the exact state?

junrao commented on core/src/main/scala/kafka/controller/ZookeeperClient.scala:
```diff
@@ -0,0 +1,214 @@
+    state.isConnected
+  }
+
+  def registerZNodeChangeHandler(zNodeChangeHandler: ZNodeChangeHandler): ExistsResponse = inReadLock(initializationLock) {
```
Do we need the read lock for all the register/unregister methods? It seems that the handle() already gets the read lock.

onurkaraman replied:
Good catch.

junrao commented on core/src/main/scala/kafka/controller/ZookeeperClient.scala:
```diff
@@ -0,0 +1,214 @@
+          isConnectedOrExpiredCondition.signalAll()
+        }
+        inWriteLock(initializationLock) {
+          sessionExpirationHandler.beforeInitializingSession
```
It seems that this block of code should only be called if (event.getState() == KeeperState.Expired)?

onurkaraman replied:
Huh. I must've somehow lost this check somewhere along the way. Good catch!

junrao commented on core/src/main/scala/kafka/controller/ZookeeperClient.scala:
```diff
@@ -0,0 +1,214 @@
+        inLock(isConnectedOrExpiredLock) {
+          isConnectedOrExpiredCondition.signalAll()
+        }
+        inWriteLock(initializationLock) {
```
The write lock probably only needs to be hold on initialize()?

onurkaraman replied:
I'm not sure. Having the lock surround the `beforeInitializingSession`, `initialize`, and `afterInitializingSession` prevents concurrent calls while these are being executed. It seemed in our very initial discussions that this is something we wanted.

junrao commented on core/src/main/scala/kafka/controller/ZookeeperClient.scala:
```diff
@@ -0,0 +1,214 @@
+  private def initialize(): Unit = {
+    if (!zooKeeper.getState.isAlive) {
+      zNodeChangeHandlers.clear()
+      zNodeChildChangeHandlers.clear()
```
It would be useful to add some info level logging to indicate that a new ZK session is created.

junrao reviewed (commented):
@onurkaraman : Thanks for the updated patch. Left a few more comments.

onurkaraman committed:
extract common attributes into AbstractRequest and AbstractResponse

onurkaraman committed:
check for expired state in the Watcher

onurkaraman committed:
remove unnecessary read locks

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6161/
Test FAILed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6145/
Test FAILed (JDK 8 and Scala 2.12).

onurkaraman committed:
just set the watch flag in handle if the watcher is set

This works because the handlers are expected to call handle to actually find out what's changed.

onurkaraman committed:
add connection timeouts

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6189/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6173/
Test PASSed (JDK 8 and Scala 2.12).

junrao commented on core/src/main/scala/kafka/controller/ZookeeperClient.scala:
```diff
@@ -0,0 +1,249 @@
+  private val zNodeChangeHandlers = new ConcurrentHashMap[String, ZNodeChangeHandler]()
+  private val zNodeChildChangeHandlers = new ConcurrentHashMap[String, ZNodeChildChangeHandler]()
+  @volatile private var zooKeeper = new ZooKeeper(connectString, sessionTimeoutMs, ZookeeperClientWatcher)
+  waitUntilConnectedOrExpired(connectionTimeoutMs, TimeUnit.MILLISECONDS)
```
We should throw an exception if waitUntilConnectedOrExpired() returns false. Also, it would be useful to indicate in the exception why the connection fails (e.g., auth failure).

onurkaraman replied:
I'm not too crazy about waitUntilConnectedOrExpired's signature returning boolean. Would it make sense to instead be:
`def waitUntilConnected(): Unit = {`
and instead have it throw potentially 3 types of exceptions:
ZookeeperClientExpiredException
ZookeeperClientAuthFailedException
ZookeeperClientTimeoutException

junrao replied:
That sounds good.

junrao commented on core/src/main/scala/kafka/controller/ZookeeperClient.scala:
```diff
@@ -0,0 +1,249 @@
+  }
+
+  def waitUntilConnectedOrExpired: Boolean = inLock(isConnectedOrExpiredLock) {
+    var state = zooKeeper.getState
```
This could now just be waitUntilConnectedOrExpired(Long.MaxLong, TimeUnit.MILLISECONDS)?

onurkaraman replied:
Makes sense. Will do.

junrao commented on core/src/main/scala/kafka/controller/ZookeeperClient.scala:
```diff
@@ -0,0 +1,249 @@
+          }
+        } catch {
+          case _: Exception =>
+            zooKeeper.close()
```
It seems that we should call zooKeeper.close() before creating a new instance. If we reach here, the new ZooKeeper object is not created.

onurkaraman replied:
Makes sense. Will do.

junrao commented on core/src/main/scala/kafka/controller/ZookeeperClient.scala:
```diff
@@ -0,0 +1,249 @@
+          inWriteLock(initializationLock) {
+            stateChangeHandler.beforeInitializingSession
+            initialize()
+            stateChangeHandler.afterInitializingSession
```
Still not sure why we can't just hold the write lock for initialize(). There is only one thread that calls beforeInitializingSession/afterInitializingSession, and zookeeper is already a volatile.

onurkaraman replied:
Suppose we only hold the write lock for initialize. Now you can permit a user to make calls concurrently as `stateChangeHandler.afterInitializingSession` is being run. I'm not sure if this is a problem.

junrao replied:
That's probably fine since Zookeeper itself is thread safe.

junrao reviewed (commented):
@onurkaraman : Thanks for the updated patch. A few more minor comments. Also, could you add some unit tests, comments, and logging?

onurkaraman committed:
close zookeeper client before creating a new instance

onurkaraman committed:
delegate unbounded waitUntilConnectedOrExpired to waitUntilConnectedOrExpired with a Long.MaxValue timeout

onurkaraman committed:
make waitUntilConnected throw exceptions when auth fails or session expirations occur instead of returning a boolean

onurkaraman committed:
add logging

onurkaraman committed:
add documentation

onurkaraman committed:
add tests for ZookeeperClient

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6286/
Test FAILed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6270/
Test FAILed (JDK 8 and Scala 2.12).

onurkaraman committed:
add license to ZookeeperClientTest

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6287/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6271/
Test PASSed (JDK 8 and Scala 2.12).

junrao commented on core/src/main/scala/kafka/controller/ZookeeperClient.scala:
```diff
@@ -0,0 +1,287 @@
+  }
+
+  def close(): Unit = inWriteLock(initializationLock) {
+    info("Closing.")
```
Should we define this.logIdent?

onurkaraman replied:
Fixed.

junrao commented on core/src/main/scala/kafka/controller/ZookeeperClient.scala:
```diff
@@ -0,0 +1,287 @@
+  private def initialize(): Unit = {
+    if (!zooKeeper.getState.isAlive) {
+      info(s"Initializing a new session to $connectString.")
+      var now = time.milliseconds()
```
Is time useful? It doesn't seem that waitUntilConnected() uses time.

onurkaraman replied:
That's right. `waitUntilConnected` doesn't use timestamps. It just uses a duration of time.

That being said, I think we agreed offline for `initialize` to try to establish a connection up until `connectionTimeoutMs` time passes. We need a timestamp in order to keep track of how much time is remaining until the connection timeout is hit.

junrao replied:
Yes, we need the timeout value. I am just not sure if we need a customized Time object. The main reason to have a Time is to mock it. Since it seems that we can't mock time in waitUntilConnected(), perhaps we don't really a Time object in the constructor of ZookeeperClient.

onurkaraman replied:
I see. Yeah that's a good point. Basically the underlying condition variable's await still uses real system time and not our mock time, so our injected time only actually helps in a subset of timing scenarios.

onurkaraman replied:
Fixed.

junrao commented on core/src/test/scala/unit/kafka/controller/ZookeeperClientTest.scala:
```diff
@@ -0,0 +1,284 @@
+    val zookeeperClient = new ZookeeperClient(zkConnect, zkSessionTimeout, zkConnectionTimeout, null, null)
+    val znodeChangeHandlerCountDownLatch = new CountDownLatch(1)
+    val zNodeChangeHandler = new ZNodeChangeHandler {
+      override def handleCreation = {}
```
Should we add a test that tests handleCreation?

junrao commented on core/src/test/scala/unit/kafka/controller/ZookeeperClientTest.scala:
```diff
@@ -0,0 +1,284 @@
+  }
+
+  @Test(expected = classOf[ZookeeperClientTimeoutException])
+  def testConnectionTimeout(): Unit = {
```
Should we add a test that fails the authentication?

junrao reviewed (commented):
@onurkaraman : Thanks for the new patch. A few more minor comments.

onurkaraman committed:
remove injected time

onurkaraman committed:
add logIdent for ZookeeperClient

onurkaraman committed:
add tests for ZNodeChangeHandler.handleCreation and StateChangeHandler.onAuthFailure

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6309/
Test FAILed (JDK 8 and Scala 2.12).

junrao commented:
@onurkaraman : Thanks for the latest patch. LGTM.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6341/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6326/
Test PASSed (JDK 8 and Scala 2.12).
